### PR TITLE
devguide: document alerts' flow - v1

### PR DESCRIPTION
**Pull request by jufajardini**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5274

Describe changes:
- changed Doxygen settings to enable call/er graph for static functions
- disabled showing unittests in call/er graphs (only `RunUnittests` shows now)
- added page about Alerts' flow
- added `Detect` as one of Suricata engines

This PR is to share what the document is looking like, to get feedback if this is the right direction or not.

### CODE REVIEW

File: doc/userguide/devguide/internals/engines/alerts-flow.rst [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+any traffic that Suricata sees.
+
+We try to be accessible to newcomers, but some level of Suricata knowledge will
+be required for the big picture to make sense.
+
+Overall Concepts
+================
+
+In Suricata, an Alert represents that a rule was triggered by a certain traffic/
+packet. Internally, we represent said matches with a ``PacketAlert`` structure
```
Comment (victorjulien) [inline doc/userguide/devguide/internals/engines/alerts-flow.rst:34]: random thought: would it make sense to link to https://doxygen.openinfosecfoundation.org/structPacketAlert__.html here?

File: doc/userguide/devguide/internals/engines/alerts-flow.rst [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+Overall Concepts
+================
+
+In Suricata, an Alert represents that a rule was triggered by a certain traffic/
+packet. Internally, we represent said matches with a ``PacketAlert`` structure 
+that is pre-processed and then appended to the Packet that triggered the
+corresponding rule.
+
+Rules are analyzed by the engine when it loads them, and given an internal id
```
Comment (victorjulien) [inline doc/userguide/devguide/internals/engines/alerts-flow.rst:38]: (`Signature::num`)

File: doc/userguide/devguide/internals/engines/alerts-flow.rst [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+packet. Internally, we represent said matches with a ``PacketAlert`` structure 
+that is pre-processed and then appended to the Packet that triggered the
+corresponding rule.
+
+Rules are analyzed by the engine when it loads them, and given an internal id 
+that symbolizes their priority. A lower id means a higher priority. Unlike 
+signature ids (``sids``), internal ids start at 0.
+
+Rule priorities are based on aspects like `action priority order
+<https://suricata.readthedocs.io/en/latest/configuration/suricata-yaml.html#suricata
```
Comment (victorjulien) [inline doc/userguide/devguide/internals/engines/alerts-flow.rst:43]: should use an sphinx link

File: doc/userguide/devguide/internals/engines/alerts-flow.rst [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    The action of an alert can be changed under specific circumstances (see
+    :doc:`/configuration/global-thresholds`). If that happens, a normal alert
+    could be converted to ``DROP``, which could affect which alerts are kept 
+    and logged.
+
+Interactions
+============
+
+Matched signatures become ``PacketAlerts`` that go to a temporary queue in
```
Comment (victorjulien) [inline doc/userguide/devguide/internals/engines/alerts-flow.rst:61]: "``PacketAlert``s"?  `PacketAlerts` != `PacketAlert`

File: doc/userguide/devguide/internals/engines/alerts-flow.rst [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+``DetectEngineThreadCtx``, where we also save the first matched `DROP` action.
+
+This queue can be dynamically expanded beyond the maximum size of the Packet's
+alert queue (``packet-alert-max``), based on the amount of matched alerts per 
+Packet. If the expansion fails, Suricata will keep working, but new alerts will
+be discarded  - ``alerts_queue_overflow`` in the stats log (see :ref:`alerts stats`).
+
+Once there are no more alerts, the queue is passed on to ``PacketAlertFinalize``,
```
Comment (victorjulien) [inline doc/userguide/devguide/internals/engines/alerts-flow.rst:69]: `PacketAlertFinalize()`? With the `()` to indicate its a function and not a struct?

File: doc/userguide/devguide/internals/engines/alerts-flow.rst [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+which will sort the queue by internal id and transaction id and then ensure 
+that the rules' action and global thresholds are respected before appending the
+alerts to the Packet.
+
+Before the queue actually being attached to the Packet, the engine will do the 
+following:
```
Comment (victorjulien) [inline doc/userguide/devguide/internals/engines/alerts-flow.rst:75]: is the order of the list relevant?

**Review by victorjulien — CHANGES_REQUESTED** [review]
Looking good. Have some minor inline comments.

File: doc/userguide/devguide/internals/engines/alerts-flow.rst (hunk 1 above)
Comment (jufajardini) [reply doc/userguide/devguide/internals/engines/alerts-flow.rst:34]: Makes sense! Gonna add it.

(on another note: Release the Kraken!!)

File: doc/userguide/devguide/internals/engines/alerts-flow.rst (hunk 6 above)
Comment (jufajardini) [reply doc/userguide/devguide/internals/engines/alerts-flow.rst:75]: I think that that is better, in order to aid the understanding of the whole thing.

**Comment by catenacyber** [comment]
Juliana, what is the status of this draft ?

**Comment by jufajardini** [comment]
> Juliana, what is the status of this draft ?

I must get back to it, I think things with higher priority got my attention and this was temporarily left aside /shame.

**Comment by jufajardini** [comment]
Closing this for now as it's still a WIP and some higher priority issues got in the way...

**Pull request closed without merging**